components/layout_2020/flexbox/layout.rs

```diff
@@ -642,13 +643,16 @@ impl FlexLine<'_> {
                 } else {
                     item_result.hypothetical_cross_size
                 };
-                let fragments = if has_stretch_auto {
-                    item.layout(used_main_size, flex_context, Some(cross_size))
-                        .fragments
-                } else {
-                    item_result.fragments
-                };
-                (cross_size, fragments)
+                if has_stretch_auto {
+                    // “If the flex item has `align-self: stretch`, redo layout for its contents,
+                    //  treating this used size as its definite cross size
+                    //  so that percentage-sized children can be resolved.”
+                    item_result = item.layout(used_main_size, flex_context, Some(cross_size));
+                }
+                flex_context
+                    .positioning_context
+                    .append(item_result.positioning_context);
```

Comment:
why was this necessary?

Reply:
as in, how does the positioning context help with side effects?

Reply:
`IndependentFormattingContext::layout` takes a `&mut PositioningContext` parameter, and has the side effect of (possibly) appending to it. To isolate this side-effect, we construct a new `PositioningContext` for each flex item, and keep it with the "result of layout". If we end up doing layout for a flex item again because of `align-self: stretch`, we throw away that `PositioningContext` and make a new one. This was added in https://github.com/servo/servo/pull/27044/commits/22b60d8b82222b542384dc8b8c2534dcfd2eda78. At some point though we want a single `PositioningContext` for the entire flex container, so we append the item-specific ones together.

I am working on a branch that removes `&mut PositioningContext` parameters everywhere and keeps hoisted boxes in respective "result of layout" types instead.